robot/control/main.cpp

```diff
@@ -72,6 +73,58 @@ KickerInfo kickerInfo;
 
 DebugInfo debugInfo;
 
+static std::vector<MODULE_META_DATA> moduleList;
+
+// Init led first to show startup progress with LED's
+static std::unique_ptr<LEDModule> led;
+
+static void superLoopTask(const void *) {
+    while (true) {
+        // Must do all timing in systick
+        // Base time we use is in systick, and if you convert from systick
+        // to us and try to compare timepoints, the value will "overflow" at
+        // some random point which messes up the `Current - Start` magic
+        // that automatically deals with delta time using the math behind overflows
+        uint32_t loopStartTime = DWT_GetTick();
+        uint32_t loopEndTime = DWT_GetTick() + SUPER_LOOP_PERIOD*DWT_SysTick_To_us();
+
+        for (unsigned int i = 0; i < moduleList.size(); i++) {
+            MODULE_META_DATA& module = moduleList.at(i);
+            uint32_t currentTime = DWT_GetTick();
+
+            // Check if we should run the module
+            //      It's time to run it and
+            //      we have enough time in the super loop
+            //
+            // Subtraction allows for rollover compensation
+            // then convertion to signed allows simple comparison
+            if (static_cast<int32_t>(currentTime - module.nextRunTime) >= 0 &&
+                static_cast<int32_t>(loopEndTime - currentTime) >= module.moduleRunTime) {
+
+                // todo change to loop start time
+                module.lastRunTime = loopStartTime;
+                module.nextRunTime = loopStartTime + module.modulePeriod;
+
+                module.module->entry();
+            }
+
+            // Check if we missed a module X times in a row
+            if (static_cast<int32_t>(currentTime - module.nextRunTime) > MAX_MISS_CNT*module.moduleRunTime) {
+                //printf("WARNING: Missed module #%d run %d times in a row\r\n", i+1, MAX_MISS_CNT);
+                led->missedModuleRun();
+            }
+        }
+
+        int32_t elapsed = DWT_GetTick() - loopStartTime;
+        if (elapsed < static_cast<int32_t>(SUPER_LOOP_PERIOD*DWT_SysTick_To_us())) {
+            DWT_Delay_Sys(SUPER_LOOP_PERIOD*DWT_SysTick_To_us() - elapsed);
+        } else {
+            //printf("WARNING: Overran super loop time\r\n");
+            led->missedSuperLoop();
+        }
+    }
+}
+
 
 int main() {
```

Comment:
You need a 'osKernalInitialize' call per https://www.keil.com/pack/doc/CMSIS/RTOS/html/group__CMSIS__RTOS__KernelCtrl.html

Reply:
ACK

Reply:
Uhh...this is actually unimplemented for our port. I'm very confused. It also doesn't seem to be necessary somehow?

Reply:
```cpp
/**
* @brief  Initialize the RTOS Kernel for creating objects.
* @retval status code that indicates the execution status of the function.
* @note   MUST REMAIN UNCHANGED: \b osKernelInitialize shall be consistent in every CMSIS-RTOS.
*/
osStatus osKernelInitialize (void);
```

I am not impressed by this CMSIS-RTOS layer.